### ISSUE TITLE
Pro::PunditIntegration: enable scoping on Arrays

### ISSUE THREAD

**Issue by rmosolgo** [rmosolgo/graphql-ruby#4726]
A while back (since #2008), I made the pundit integration skip scoping on Arrays because Pundit didn't handle Arrays nicely. 

As pointed out in https://github.com/rmosolgo/graphql-ruby/pull/3994#issuecomment-1711539486 and https://github.com/rmosolgo/graphql-ruby/pull/3994#issuecomment-1838977651, this can lead to unexpectedly _skipping_ authorization when field return values switch from ActiveRecord::Relations to plain Arrays. (For example, if you switch to loading records by GraphQL::Dataloader.)

There _is_ a way to opt back in to scoping arrays with Pundit, described in https://github.com/rmosolgo/graphql-ruby/issues/3173#issuecomment-702736339. I think I'm going to change the default back to that behavior. 

That means: 

- If a field returns an array, 
- And that field is scoped, 
- And the return type of that field doesn't have a manually-configured `pundit_policy_class`, 
- Then it will raise a weird error for failing to find a scope for an array. 

I have wrapped that error to include some debug info and suggestions: 

```
No policy found for:

- Type: Beast
- Runtime value: `[#<Beast id: 1, name: "Sasquatch", nickname: "Big Foot", current_whereabouts_description: "The Great White North">, #<Beast id: 2, name: "Loch Ness Monster", nickname: "Nessie", current_whereabouts_description: "Depths of Loch Ness">]`

Since this value is an **Array**, Pundit can't find a scope for it. Configure one with `pundit_policy_class` in the definition for `Beast`.
Or, you can skip scoping by adding `scope: false` to the field definition or implementing `Beast.scope_items(items, ctx)` to return `items` without calling `super(...)`.

Pundit error: Pundit::NotDefinedError, unable to find scope `Beast::Beast::Scope` for `[#<Beast id: 1, name: "Sasquatch", nickname: "Big Foot", current_whereabouts_description: "The Great White North">, #<Beast id: 2, name: "Loch Ness Monster", nickname: "Nessie", current_whereabouts_description: "Depths of Loch Ness">]`
```

In this approach, you can opt back out using `scope: false` on the field or implementing `def self.scope_items` _not_ to call `super(...)` (as also mentioned in the error message). 

This will be a "breaking" change in that any fields which meet the criteria described above will _start_ raising errors. But assuming the field is used in a test _somewhere_, it should be easy to find and easy to fix. 

Any thoughts on this proposal?

**Comment by rmosolgo** [comment]
@bessey I'd love your thoughts on this if you have a minute to consider it!

**Comment by bessey** [comment]
Its a hectic time at work just now, but I will get to this eventually

**Comment by bessey** [comment]
I like this proposal @rmosolgo. 

Perhaps this could be made a less scary change by being configurable at the schema level to 
log a deprecation warning rather than raise?

Speaking for my org, we probably do have the E2E testing in place to be confident going straight to raising.

**Comment by rmosolgo** [comment]
:+1: I just released graphql-pro v1.26.0 with this change. I decided to stick with raising an error -- I figure it's worth making this change obvious, and I expect most people, like you, will have some tests that hit these code paths.
